examples/drupal_loadtest.rs

```diff
@@ -85,39 +85,41 @@ async fn drupal_loadtest_front_page(user: &GooseUser) {
     let mut goose = user.get("/").await;
 
     // Grab some static assets from the front page.
-    match goose.response {
-        Ok(r) => {
-            // Copy the headers so we have them for logging if there are errors.
-            let headers = &r.headers().clone();
-            match r.text().await {
-                Ok(t) => {
-                    let re = Regex::new(r#"src="(.*?)""#).unwrap();
-                    // Collect copy of URLs to run them async
-                    let mut urls = Vec::new();
-                    for url in re.captures_iter(&t) {
-                        if url[1].contains("/misc") || url[1].contains("/themes") {
-                            urls.push(url[1].to_string());
+    if let Some(response) = goose.response {
```

Comment:
Maybe an early return (as we know that is the error case to exit on) would help to remove this much indentation changes and so many levels deep in indentation?

While I am not a fan of it necessarily, an Exception usually is thrown in other languages when the client is no longer valid (as the load test should end). This Exception is then caught by the thread runner which cleanly ends the loadtest.

Even without Exceptions an early return should be what is done here - as all cleanup if needed should be in a different task and not here.

Reply:
I'm confused what we're accomplishing. We'd still have to unwrap the `goose.response` to see if it's None or `Some()`. Your goal is simply to minimize indention?

Reply:
Okay - I've informed myself:

The thing to wrap in an `Result<T, Goose>` is the return of the .get(), etc. functions itself.

As load test ended is an Err on the Goose, not on reqwest.

Then the question mark operator would work as well:

```
    let mut goose = user.get("/").await?;

    // Grab some static assets from the front page.
    match goose.response {
```

is replaced with:

```
    // Early return on error
    let mut goose = match user.get("/").await {
           Err(e) => return Err(e),
           Ok(f) => f,
    };

    // Grab some static assets from the front page.
    match goose.response {
```

That is also way more rust-y :).

Reference:

https://static.rust-lang.org/doc/master/std/result/index.html

Reply:
Here's one way to exit early, but I'm not convinced this is better ...?  The end result is the same, but this is actually a little extra code and the addition of a `return` which feels kind of like using a `goto`.

```rust
diff --git a/examples/drupal_loadtest.rs b/examples/drupal_loadtest.rs
index 1ad89b2..266e97f 100644
--- a/examples/drupal_loadtest.rs
+++ b/examples/drupal_loadtest.rs
@@ -84,42 +84,47 @@ fn main() {
 async fn drupal_loadtest_front_page(user: &GooseUser) {
     let mut goose = user.get("/").await;
 
+    let response = match goose.response {
+        // Unwrap response.
+        Some(r) => (r),
+        // If there was no response, exit.
+        None => return,
+    };
+
     // Grab some static assets from the front page.
-    if let Some(response) = goose.response {
-        match response {
-            Ok(r) => {
-                // Copy the headers so we have them for logging if there are errors.
-                let headers = &r.headers().clone();
-                match r.text().await {
-                    Ok(t) => {
-                        let re = Regex::new(r#"src="(.*?)""#).unwrap();
-                        // Collect copy of URLs to run them async
-                        let mut urls = Vec::new();
-                        for url in re.captures_iter(&t) {
-                            if url[1].contains("/misc") || url[1].contains("/themes") {
-                                urls.push(url[1].to_string());
-                            }
-                        }
-                        for asset in &urls {
-                            user.get_named(asset, "static asset").await;
+    match response {
+        Ok(r) => {
+            // Copy the headers so we have them for logging if there are errors.
+            let headers = &r.headers().clone();
+            match r.text().await {
+                Ok(t) => {
+                    let re = Regex::new(r#"src="(.*?)""#).unwrap();
+                    // Collect copy of URLs to run them async
+                    let mut urls = Vec::new();
+                    for url in re.captures_iter(&t) {
+                        if url[1].contains("/misc") || url[1].contains("/themes") {
+                            urls.push(url[1].to_string());
                         }
                     }
-                    Err(e) => {
-                        user.set_failure(&mut goose.request);
-                        let error = format!("front_page: failed to parse pag: {}", e);
-                        // We choose to both log and display errors to stdout.
-                        eprintln!("{}", &error);
-                        user.log_debug(&error, Some(goose.request), Some(&headers), None);
+                    for asset in &urls {
+                        user.get_named(asset, "static asset").await;
                     }
                 }
+                Err(e) => {
+                    user.set_failure(&mut goose.request);
+                    let error = format!("front_page: failed to parse pag: {}", e);
+                    // We choose to both log and display errors to stdout.
+                    eprintln!("{}", &error);
+                    user.log_debug(&error, Some(goose.request), Some(&headers), None);
+                }
             }
-            Err(e) => {
-                user.set_failure(&mut goose.request);
-                let error = format!("front_page: no response from server: {}", e);
-                // We choose to both log and display errors to stdout.
-                eprintln!("{}", &error);
-                user.log_debug(&error, Some(goose.request), None, None);
-            }
+        }
+        Err(e) => {
+            user.set_failure(&mut goose.request);
+            let error = format!("front_page: no response from server: {}", e);
+            // We choose to both log and display errors to stdout.
+            eprintln!("{}", &error);
+            user.log_debug(&error, Some(goose.request), None, None);
         }
     }
 }
```

Reply:
And yes - we need to change our function signature again for the load test functions, but it's cleaner to return a result anyway.

Even though we need to add a Ok(()) at the end, that's still cleaner than these if chains.

Also while I am unclear how that works in practice, it should be possible to also return an error from Reqwest early that way - though we probably need to provide an `into trait`, which converts one type in another.

Reply:
The other option is to change the task function signature, requiring them to return a [result](https://doc.rust-lang.org/std/result/).

I'm onboard with the general idea, but it's a much larger change and would need to be done in a different PR. It will need further thought/discussion/architecture to decide how Goose handles errors. (ie, do we increment a statistic somewhere? is it a different statistic depending on the error? Can we auto-magically implement `set_failure()` and `set_success()`?) I'd prefer to explore this as a possible followup.

Reply:
Ok, reworked to return a `Result` instead of making `goose.response` optional.

This begs a followup to let task functions return a Result as well, but that will happen in a followup.

Reply:
Further reworked to exit early, setting us up for using `?` in the load test function itself when supported.